RABBITHOLE.md

```diff
@@ -0,0 +1,93 @@
+
+* I've yet to run into a place where type-based registration is problematic.
+* That said, service-based listeners must have their type specified explicilty; the reflection detection, despite being fun to say, doesn't work without an for-reals callable (obviously).  I don't see that as a serious drawback.
+* The service-based listener use case obviously needs a container, and PSR-11 takes care of everything we need.  Which... suggests PSR-11 may have been a good thing after all. :-)
+* The compiled use-case is handled best not by a compiled dispatcher but a compiled ListenerSet.
+* In fact, with listener sets (aggregators, whatever) factored out to a separate interface a nearly universal dispatcher is stupid-easy to write.  The only use case for an alternate implementation seems to be for instrumentation (eg, logging every event directly for debugging).  I consider this a good thing.
+* However, there seems little value to a unified registration interface.  Every use case (unordered, priority ordered, before/after ordered, services, etc.) seems to require a different method signature.  That suggests we should do one of the following:
+** Do not provide a registration interface at all, and explicitly kick that downstream to implementers.  This has the benefit of flexibility but the downside of having no standardized way for trivial cases to register themselves.
+** Provide multiple "common case" registration interfaces (probably unordered and priority ordered) and then explicitly say that they're optional and it's totally legit to define your own instead.
+* The word "listener" is far too easy to misspell as "listner".
+* The code in pretty much all implementations is scarily simple.  That is a good sign.
+
+## The interfaces
+
+```php
+interface EventInterface
+{
+    public function stopPropagation(bool $stop = true) : self;
+
+    public function stopped() : bool;
+}
+```
+
+A trivially simple interface with propagation control.  There's also a trait provided that implements them both in the only logical way, which is a natural fit for a util library.
+
+```php
+interface ListenerSetInterface
+{
+    public function getListenersFor(EventInterface $event): iterable;
+}
+```
+
+This is a read-only collection of listeners.  It's basically the same as Matthew's "aggregate", as that turned out to be a really elegant solution but I find "aggregate" to be too pretentious a word even for me :-).  The only thing you can do is get a list of listeners that are applicable to an event object.  It returns an iterable so that implementers can use an array or generator or whatever at their leisure.
+
+```php
+interface DispatcherInterface
+{
+    public function dispatch(EventInterface $event) : EventInterface;
```

Comment:
I need to check internally to see if forcing the event "type" to be the event class will cause problems for us.

Reply:
My expectation is that Symfony 4 will support both styles simultaneously and Symfony 5 would go full-FIG.